add compass ui in common components package

AshishDhama opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does, as well as QA test steps (if applicable and if not already added to the Jira ticket).
-->

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates).
* API additions—new endpoint, new response fields, or newly accepted request parameters.
* Database changes (any).
* Schema migration changes. Use the [Schema Migration Template](https://docs.google.com/document/d/18lD7N32oyMtYjFrJKwsNv8yn6Fe5QtF-eMm8nn0O8tk/edit?usp=sharing) as a starting point to capture these details as release notes. 
* Websocket additions or changes.
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating).
* New features and improvements, including behavioral changes, UI changes, and CLI changes.
* Bug fixes and fixes of previous known issues.
* Deprecation warnings, breaking changes, or compatibility notes.

If no release notes are required, write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note

```

nevyangelova commented on webapp/platform/components/src/compass-ui/components/base_modal/modal_title.tsx:
```diff
@@ -0,0 +1,71 @@
+    padding: hasCloseButton ? '22px 26px 24px 32px' : '28px 32px 24px',
+}));
+
+const StyledModalTitleSection = styled('div')(() => ({
```
we don't need to pass a function in the instances where there aren't props. 
```
styled.div`...styles`;
```

nevyangelova commented on webapp/channels/src/components/new_channel_modal/new_channel_modal.test.tsx:
```diff
@@ -388,57 +385,58 @@ describe('components/new_channel_modal', () => {
-            update_at: 0,
-        }, '');
-    });
+    // disable test temporarily
```
should we merge without this test?

AshishDhama replied:
I have updated the test.

nevyangelova commented on webapp/platform/components/src/compass-ui/components/base_modal/base_modal.tsx:
```diff
@@ -0,0 +1,64 @@
+    },
+    ref: React.Ref<unknown>,
+) => {
+    return (
```
nit: shorthand return

nevyangelova commented on webapp/platform/components/src/compass-ui/components/button/button.overrides.ts:
```diff
@@ -0,0 +1,173 @@
+    },
+});
+
+const styleOverrides: ComponentsOverrides<Theme>[typeof componentName] = {
```
just to confirm. The overrides are taken from the Los Tigres discussions right?

AshishDhama replied:
Yeah

nevyangelova commented on webapp/platform/components/src/compass-ui/components/button/button.overrides.ts:
```diff
@@ -0,0 +1,173 @@
+};
+
+const getFocusStyles = (color: string) => ({
+    '&:not(.Mui-disabled)': {
```
we can wrap the whole style block in `` marks, I am not sure if they are applied correctly this way?

nevyangelova commented on webapp/platform/components/src/compass-ui/components/icon_button/icon_button.overrides.ts:
```diff
@@ -0,0 +1,206 @@
+};
+
+export const getFocusStyles = (color: string) => ({
+    '&:not(.Mui-disabled)': {
```
same here, can we wrap the css blocks together?

nevyangelova commented on webapp/platform/components/src/compass-ui/components/modal/generic_modal.tsx:
```diff
@@ -0,0 +1,80 @@
+                {children}
+                {hasActions && (
+                    <DialogActions>
+                        {onCancel && <Button onClick={handleCancelAction}>{'Cancel'}</Button>}
```
assuming these will be translated?

AshishDhama replied:
@hmhealey any suggestions for this?

hmhealey replied:
You're able to use react-intl here, but i18n-extract isn't currently able to find the translations on its own, so you'd need to add a `t` function somewhere in `channels` with this string. We've already done that for the current `GenericModal` [here](https://github.com/mattermost/mattermost-server/blob/master/webapp/channels/src/components/generic_modal.tsx#L10-L16), so you can probably just reuse those same IDs here

nevyangelova commented on webapp/platform/components/src/compass-ui/index.ts:
```diff
@@ -0,0 +1,12 @@
+
+export * from './components';
+
+// import ThemeProvider from './themeprovider/themeprovider';
```
todo: add theme provider per https://mattermost.atlassian.net/browse/MM-52136

nevyangelova reviewed (commented):
Thanks @AshishDhama some minor suggestions. Looks good!

hmhealey commented on webapp/platform/components/src/compass-ui/themeprovider/themes.ts:
```diff
@@ -0,0 +1,89 @@
+    }
+}
+
+export const lightTheme = createTheme({
```
What are these light and dark themes needed for?

hmhealey commented on webapp/platform/components/src/compass-ui/themeprovider/themeprovider.tsx:
```diff
@@ -0,0 +1,24 @@
+import {lightTheme} from './themes';
+import overrides from './overrides';
+
+const ThemeProvider = ({theme = lightTheme, ...rest}) => {
```
Just a thought, but we could make this take a MM theme and automatically convert it into a MUI one rather than making the person using the ThemeProvider do that conversion itself. That'd be nice since it would further abstract away the MUI parts of this

hmhealey commented on webapp/platform/components/src/compass-ui/utils/color-utils.ts:
```diff
@@ -0,0 +1,32 @@
+// See LICENSE.txt for license information.
+
+import {decomposeColor, hslToRgb} from '@mui/material';
+import {normal} from 'color-blend';
```
Could you add this library to the dependencies of this package? We're currently getting it from the web app importing `compass-components`, so it'd be better to be explicit about it as a dependency

hmhealey commented on webapp/platform/components/src/index.tsx:
```diff
@@ -2,13 +2,14 @@
 // components
-export * from './generic_modal/footer_content';
-export {GenericModal} from './generic_modal/generic_modal';
+export * from './compass-ui';
```
Nothing to change here for now, but this is already starting to make me question exporting all of these components as a single module. The module's not particularly big yet, but it doesn't support code splitting, so it might start to cause the bundle size to grow a lot once more components are added

hmhealey reviewed (changes requested):
I have a few comments/requests for the structure of these files:
1. A few of these files/folders don't follow our usual naming convention. Would you mind renaming `themeprovider` to `theme_provider`, `color-utils` to `color_utils` (or even just `utils/color`), and the same thing for `font-utils`?
2. Similarly but more nitpicky, could we rename the `compass-ui` folder to just `compass`? The UI part seems unnecessary
3. Why not have an index.ts file in component folders like `components/base_modal`?

AshishDhama commented on webapp/platform/components/src/compass-ui/components/icon_button/icon_button.tsx:
```diff
@@ -0,0 +1,51 @@
+            className={classNames({compact, toggled, inverted})}
+            color={destructive ? 'error' : 'primary'}
+        >
+            <Grid
```
Yeah, But I think we might need make some adjustment in future

AshishDhama commented:
@hmhealey I have taken care of most of your feedback, There are few things which I have left which I would like to take care in next pr as we have already reached 41 files. 
2 Things I have done, moved couple theme types to types package but I haven’t removed them from redux store package,
Moved some constants to components package I am not sure where to put things which are required in channels and components both, to avoid circular dependency situation.
I really appreciate the detailed feedback.

AshishDhama commented:
/update-branch

hmhealey commented on webapp/platform/types/src/theme.ts:
```diff
@@ -0,0 +1,35 @@
+export type ThemeType = 'Denim' | 'Sapphire' | 'Quartz' | 'Indigo' | 'Onyx';
+
+export type Theme = {
+    [key: string]: string | undefined;
```
If this doesn't cause any type issues, would you mind removing this? I don't 100% remember why we made it so that you could add arbitrary values to the Theme.

If it's because of something we add in `utils/constants` versions of the themes, let's remove that from this version and we can fix it when we go to use this type in Channels

hmhealey commented on webapp/platform/types/src/theme.ts:
```diff
@@ -0,0 +1,35 @@
+    awayIndicator: string;
+    dndIndicator: string;
+    mentionBg: string;
+    mentionBj: string;
```
Similarly, could you remove this one as well? It's the result of an ancient typo so I don't want to copy it over to the new package

AshishDhama replied:
<img width="524" alt="Screenshot 2023-05-03 at 11 49 20 AM" src="https://user-images.githubusercontent.com/16203333/235844351-870d6631-eb8e-4c75-9e0d-5a72570a62cb.png">
What about this?

AshishDhama replied:
Although I am in favour of removing it in 8.0 or at-least depreciation warning

hmhealey replied:
For that one, I think we can have it use `(theme as any).mentionBj`. It's kind of ugly, but it's the only place that the old field name should ever be needed

hmhealey commented on webapp/platform/components/src/compass/components/button/button.overrides.ts:
```diff
@@ -0,0 +1,173 @@
+    },
+});
+
+const styleOverrides: ComponentsOverrides<Theme>[typeof componentName] = {
```
I think my previous comment here got lost in all the renaming (thanks for doing that by the way!):

Is this something we plan to keep in the longer term? It seems weird to me that we're having to inject some overrides into the Provider instead of doing these overrides on the component itself

AshishDhama replied:
I will discuss this with Nevy, Mostly we will do if something can be done in https://github.com/mattermost/mattermost-server/pull/23238.

hmhealey commented on webapp/platform/components/src/compass/components/generic_modal/generic_modal.tsx:
```diff
@@ -0,0 +1,80 @@
+                {children}
+                {hasActions && (
+                    <DialogActions>
+                        {onCancel && <Button onClick={handleCancelAction}>{'Cancel'}</Button>}
+                        {onConfirm && <Button onClick={handleConfirmAction}>{'Confirm'}</Button>}
```
Same with my comment here. We can use `react-intl` here as long as we use a `t` function somewhere in `webapp/channels` to make sure the i18n tooling knows the string exists. In this case, we can just reuse the strings defined for the LegacyGenericModal [here](https://github.com/mattermost/mattermost-server/blob/master/webapp/channels/src/components/generic_modal.tsx#L10-L16)

AshishDhama replied:
Yeah sure, This pr was meant to introduce compass package. I will add these changes in https://github.com/mattermost/mattermost-server/pull/23238.

AshishDhama commented:
/update-branch

AshishDhama commented:
/e2e-test

mattermost-build commented:
Successfully triggered E2E testing!
[GitLab pipeline](https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/332543) | [Test dashboard](https://automation-dashboard.vercel.app/cycle/332543)
